File: src/services/SpotifyService.ts
import { fetch as tFetch } from "@tauri-apps/plugin-http";

const DEFAULT_CLIENT_ID = "b1c574848d0b491eb75f94f515e9c7de";
const DEFAULT_CLIENT_SECRET = "e5593f4ca9644a4c8ea03ec0b3178913";
const FETCH_TIMEOUT_MS = 8000;

let cachedToken: string | null = null;
let tokenExpiry = 0;
let cachedCredKey = "";

function fetchWithTimeout(url: string, options: Parameters<typeof tFetch>[1] = {}, ms = FETCH_TIMEOUT_MS) {
  const abort = new AbortController();
  const timer = setTimeout(() => abort.abort(), ms);
  return tFetch(url, { ...options, signal: abort.signal }).finally(() => clearTimeout(timer));
}

function getSpotifyCreds() {
  const id = localStorage.getItem("tagwave_spotify_id") || DEFAULT_CLIENT_ID;
  const secret = localStorage.getItem("tagwave_spotify_secret") || DEFAULT_CLIENT_SECRET;
  return { id, secret };
}

async function getToken(): Promise<string> {
  const { id, secret } = getSpotifyCreds();
  const credKey = id + ":" + secret;
  if (credKey !== cachedCredKey) { cachedToken = null; tokenExpiry = 0; cachedCredKey = credKey; }
  if (cachedToken && Date.now() < tokenExpiry) return cachedToken!;
  const creds = btoa(credKey);
  const res = await fetchWithTimeout("https://accounts.spotify.com/api/token", {
    method: "POST",
    headers: { Authorization: `Basic ${creds}`, "Content-Type": "application/x-www-form-urlencoded" },
    body: "grant_type=client_credentials",
  });
  const data = await res.json();
  cachedToken = data.access_token;
  tokenExpiry = Date.now() + (data.expires_in - 60) * 1000;
  return cachedToken!;
}

const NOTES = ["C","C#","D","D#","E","F","F#","G","G#","A","A#","B"];
const CAMELOT_MAJOR = ["8B","3B","10B","5B","12B","7B","2B","9B","4B","11B","6B","1B"];
const CAMELOT_MINOR = ["5A","12A","7A","2A","9A","4A","11A","6A","1A","8A","3A","10A"];

export interface SpotifyFeatures {
  bpm: string;
  key: string;
  camelot: string;
  energy: number;
  danceability: number;
}

export interface SpotifyInfo {
  album: string;
  year: string;
  coverUrl: string;
  features: SpotifyFeatures | null;
  matchPhase?: "exact" | "fuzzy";
}

const VERSION_PATTERN = /\s*[\[(](extended|radio|club|original|dub|instrumental|remix|mix|edit|version|vocal|acapella|live|acoustic|reprise|remaster\w*|feat\.?|ft\.?)(\s+[^\])]*)?[\])]\s*$/gi;
const VERSION_DASH    = /\s+-\s+(extended|radio|club|original|dub|instrumental|remix|mix|edit|version|vocal|live|acoustic)\s*$/gi;

function stripVersionInfo(title: string): string {
  return title.replace(VERSION_PATTERN, "").replace(VERSION_DASH, "").trim();
}

async function querySpotify(term: string, artist: string, token: string): Promise<{ id: string; album: string; year: string; coverUrl: string } | null> {
  const clean = term.replace(/\.(mp3|flac|wav|aiff?)$/i, "").trim();
  const q = artist ? `${clean} artist:${artist}` : clean;
  const res = await fetchWithTimeout(
    `https://api.spotify.com/v1/search?q=${encodeURIComponent(q)}&type=track&limit=1`,
    { headers: { Authorization: `Bearer ${token}` } }
  );
  const data = await res.json();
  const item = data.tracks?.items?.[0];
  if (!item) return null;
  return {
    id: item.id,
    album: item.album?.name ?? "",
    year: (item.album?.release_date ?? "").slice(0, 4),
    coverUrl: item.album?.images?.[0]?.url ?? "",
  };
}

async function searchTrackId(
  title: string,
  artist: string,
  token: string,
): Promise<{ id: string; album: string; year: string; coverUrl: string; matchPhase: "exact" | "fuzzy" } | null> {
  // Fase 1 — nome completo (com versão)
  const hit1 = await querySpotify(title, artist, token);
  if (hit1) return { ...hit1, matchPhase: "exact" };

  // Fase 2 — sem info de versão
  const stripped = stripVersionInfo(title);
  if (stripped && stripped !== title) {
    const hit2 = await querySpotify(stripped, artist, token);
    if (hit2) return { ...hit2, matchPhase: "fuzzy" };
  }

  return null;
}

async function getAudioFeatures(trackId: string, token: string): Promise<SpotifyFeatures | null> {
  try {
    const res = await fetchWithTimeout(`https://api.spotify.com/v1/audio-features/${trackId}`, {
      headers: { Authorization: `Bearer ${token}` },
    });
    if (res.status === 403) return null;
    const f = await res.json();
    if (typeof f.key !== "number" || f.key < 0) return null;
    return {
      bpm: String(Math.round(f.tempo)),
      key: f.mode === 1 ? NOTES[f.key] : `${NOTES[f.key]}m`,
      camelot: f.mode === 1 ? CAMELOT_MAJOR[f.key] : CAMELOT_MINOR[f.key],
      energy: f.energy ?? 0,
      danceability: f.danceability ?? 0,
    };
  } catch {
    return null;
  }
}

export async function enrichTrackFull(title: string, artist: string): Promise<SpotifyInfo | null> {
  try {
    const token = await getToken();
    const track = await searchTrackId(title, artist, token);
    if (!track) return null;
    const features = await getAudioFeatures(track.id, token);
    return { album: track.album, year: track.year, coverUrl: track.coverUrl, features, matchPhase: track.matchPhase };
  } catch {
    return null;
  }
}

export async function enrichTrack(title: string, artist: string): Promise<SpotifyFeatures | null> {
  const info = await enrichTrackFull(title, artist);
  return info?.features ?? null;
}

export async function batchEnrich(
  tracks: Array<{ id: string; title: string; artist: string; bpm?: string }>,
  onProgress: (id: string, info: SpotifyInfo) => void
): Promise<void> {
  for (const t of tracks) {
    const info = await enrichTrackFull(t.title || t.artist, t.artist);
    if (info) onProgress(t.id, info);
  }
}

File: swift-macos/Sources/MP3Manager/Services/SpotifyService.swift
import Foundation

struct SpotifyAudioFeatures {
    let bpm: String
    let key: String        // Notação padrão (ex: "C", "Cm", "F#", "Bbm")
    let camelotKey: String // Notação Camelot (ex: "8B", "5A") — para referência
    let energy: Double
    let danceability: Double
    let valence: Double
}

struct SpotifyTrackInfo {
    var album: String = ""
    var year: String = ""
    var coverURL: URL? = nil
    var audioFeatures: SpotifyAudioFeatures? = nil // nil se endpoint deprecado
}

enum SpotifyError: Error {
    case notFound
    case audioFeaturesDeprecated  // 403 — endpoint removido para apps novos
    case httpError(Int)
    case decodingError
}

final class SpotifyService {
    static let shared = SpotifyService()
    private init() {}

    private let clientId     = "b1c574848d0b491eb75f94f515e9c7de"
    private let clientSecret = "e5593f4ca9644a4c8ea03ec0b3178913"
    private var accessToken: String?
    private var tokenExpiry: Date = .distantPast
    private let session = URLSession.shared

    // MARK: - Public

    /// Enriquece uma faixa com dados do Spotify.
    /// Sempre tenta buscar álbum/ano/capa. BPM/Tom apenas se audio-features disponível.
    func enrich(_ track: Track) async -> SpotifyTrackInfo? {
        guard !track.title.isEmpty else { return nil }
        do {
            let token = try await getToken()
            guard let (id, info) = try await searchTrack(title: track.title, artist: track.artist, token: token) else { return nil }

            var result = info

            // Tenta audio-features (pode falhar com 403 em apps criados após Nov 2024)
            if let features = try? await getAudioFeatures(id: id, token: token) {
                result.audioFeatures = features
            }

            return result
        } catch { return nil }
    }

    func batchEnrich(tracks: [Track], appState: AppState) async {
        await MainActor.run {
            appState.isSpotifyEnriching   = true
            appState.spotifyEnrichDone    = 0
            appState.spotifyEnrichTotal   = tracks.count
            appState.spotifyEnrichCurrent = ""
        }

        var enrichedCount = 0

        for track in tracks {
            if Task.isCancelled { break }
            await MainActor.run { appState.spotifyEnrichCurrent = track.url.path }

            if let info = await enrich(track) {
                var updated = track
                // BPM e Tom — só se audio-features disponível
                if let f = info.audioFeatures {
                    updated.bpm = f.bpm
                    updated.key = f.key
                }
                // Álbum e ano — sempre aplica se vieram vazios
                if updated.album.isEmpty && !info.album.isEmpty { updated.album = info.album }
                if updated.year.isEmpty  && !info.year.isEmpty  { updated.year  = info.year }

                ValidationService.revalidate(&updated)
                let toWrite = updated
                await MainActor.run { appState.updateTrack(toWrite) }
                try? await TagWriter.shared.writeTags(to: toWrite)
                enrichedCount += 1
            }

            await MainActor.run { appState.spotifyEnrichDone += 1 }
            try? await Task.sleep(nanoseconds: 120_000_000)
        }

        let n = enrichedCount
        await MainActor.run {
            appState.isSpotifyEnriching = false
            appState.spotifyEnrichTask  = nil
            if n == 0 {
                appState.statusMessage = "Spotify: nenhuma faixa encontrada (verifique título/artista)"
            } else {
                appState.statusMessage = "✓ Spotify: \(n) faixa\(n == 1 ? "" : "s") atualizada\(n == 1 ? "" : "s")"
            }
        }
    }

    // MARK: - Auth

    private func getToken() async throws -> String {
        if let t = accessToken, Date() < tokenExpiry { return t }

        var req = URLRequest(url: URL(string: "https://accounts.spotify.com/api/token")!)
        req.httpMethod = "POST"
        let creds = Data("\(clientId):\(clientSecret)".utf8).base64EncodedString()
        req.setValue("Basic \(creds)", forHTTPHeaderField: "Authorization")
        req.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        req.httpBody = Data("grant_type=client_credentials".utf8)

        let (data, _) = try await session.data(for: req)
        let json = try JSONDecoder().decode(TokenResponse.self, from: data)
        accessToken = json.access_token
        tokenExpiry = Date().addingTimeInterval(Double(json.expires_in) - 60)
        return json.access_token
    }

    // MARK: - Search + Track info

    /// Busca a faixa e retorna (spotifyId, TrackInfo com álbum/ano/capa)
    private func searchTrack(title: String, artist: String, token: String) async throws -> (String, SpotifyTrackInfo)? {
        let cleanTitle = title.replacingOccurrences(of: #"\.(mp3|flac|wav|aiff?)$"#, with: "", options: .regularExpression)
        let q = artist.isEmpty ? cleanTitle : "\(cleanTitle) artist:\(artist)"

        var comps = URLComponents(string: "https://api.spotify.com/v1/search")!
        comps.queryItems = [
            .init(name: "q",     value: q),
            .init(name: "type",  value: "track"),
            .init(name: "limit", value: "1"),
        ]
        var req = URLRequest(url: comps.url!)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, resp) = try await session.data(for: req)
        if let http = resp as? HTTPURLResponse, http.statusCode != 200 { return nil }

        let json = try JSONDecoder().decode(SearchResponse.self, from: data)
        guard let item = json.tracks.items.first else { return nil }

        var info = SpotifyTrackInfo()
        info.album = item.album.name
        // release_date pode ser "2001", "2001-05", "2001-05-21"
        info.year  = String(item.album.release_date.prefix(4))
        if let imgUrl = item.album.images.first?.url {
            info.coverURL = URL(string: imgUrl)
        }

        return (item.id, info)
    }

    // MARK: - Audio Features (pode retornar 403 em apps novos)

    private func getAudioFeatures(id: String, token: String) async throws -> SpotifyAudioFeatures? {
        var req = URLRequest(url: URL(string: "https://api.spotify.com/v1/audio-features/\(id)")!)
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, resp) = try await session.data(for: req)
        if let http = resp as? HTTPURLResponse {
            if http.statusCode == 403 { throw SpotifyError.audioFeaturesDeprecated }
            if http.statusCode != 200 { throw SpotifyError.httpError(http.statusCode) }
        }

        let f = try JSONDecoder().decode(AudioFeatures.self, from: data)
        guard f.key >= 0 else { return nil }
        return SpotifyAudioFeatures(
            bpm:          String(Int(f.tempo.rounded())),
            key:          standardKey(key: f.key, mode: f.mode),
            camelotKey:   camelotKey(key: f.key, mode: f.mode),
            energy:       f.energy,
            danceability: f.danceability,
            valence:      f.valence
        )
    }

    // MARK: - Key notation

    private func standardKey(key: Int, mode: Int) -> String {
        let notes = ["C","C#","D","D#","E","F","F#","G","G#","A","A#","B"]
        guard key < 12 else { return "" }
        return mode == 1 ? notes[key] : "\(notes[key])m"
    }

    private func camelotKey(key: Int, mode: Int) -> String {
        let major = ["8B","3B","10B","5B","12B","7B","2B","9B","4B","11B","6B","1B"]
        let minor = ["5A","12A","7A","2A","9A","4A","11A","6A","1A","8A","3A","10A"]
        guard key < 12 else { return "" }
        return mode == 1 ? major[key] : minor[key]
    }

    // MARK: - Decodable models

    private struct TokenResponse: Decodable {
        let access_token: String
        let expires_in: Int
    }

    private struct SearchResponse: Decodable {
        struct Tracks: Decodable {
            struct Item: Decodable {
                let id: String
                let album: Album
            }
            struct Album: Decodable {
                let name: String
                let release_date: String
                let images: [SpotifyImage]
            }
            struct SpotifyImage: Decodable {
                let url: String
            }
            let items: [Item]
        }
        let tracks: Tracks
    }

    private struct AudioFeatures: Decodable {
        let tempo: Double
        let key: Int
        let mode: Int
        let energy: Double
        let danceability: Double
        let valence: Double
    }
}
